### PR TITLE
Introduce constrain_range; remove old expr_subs

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95063

This PR introduces a new `constrain_range` function which can be used to constrain the possible values a SymInt/SymFloat can take on. This knowledge can be then used to discharge potential guards (by running the range analysis, and then seeing if the guard must be true given the original range) without adding another guard.

The usage of ranges is very limited right now; ranges are only constrained when the user explicitly instructs the system so. However, we can also infer range constraints based on guards as well; this is left for future work.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -116,6 +118,23 @@ def guard_scalar(a):
     else:
         raise AssertionError(f"unrecognized scalar {a}")
 
+# inclusive both ways
+def constrain_range(a, *, min, max=sympy.oo):
+    if not isinstance(a, SymInt):
+        assert min <= a <= max
+        return
+    if isinstance(a.node.expr, sympy.Integer):
```
Comment (eellison) [inline torch/fx/experimental/symbolic_shapes.py:126]: any reason not to support Floats here ?

**Review by eellison — APPROVED** [review]
Nice. We could start producing a range for Symints like `.nonzero()` at some point.

File: torch/utils/_sympy/interp.py [hunk 2]
```diff
@@ -66,7 +66,7 @@ def sympy_interp(
     # sometimes?
     if isinstance(expr, sympy.Integer):
         return analysis.constant(int(expr), torch.int64)
-    elif isinstance(expr, sympy.Float):
+    elif isinstance(expr, sympy.Number):
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/utils/_sympy/interp.py:69]: 1/3 turns into a Rational which is not a float. There's something kind of fishy going on here... I am still feeling like I want to pass the sympy expressions directly to analysis.constant without casting it to float first.

File: torch/fx/experimental/symbolic_shapes.py [hunk 3]
```diff
@@ -1070,6 +1089,11 @@ def __init__(self, allow_scalar_outputs=True):
         # Maps symbolic ints to their original concrete values
         # Currently populated from tensors
         self.var_to_val: Dict["sympy.Symbol", "sympy.Integer"] = {}
+        # Maps symbolic ints to their min/max range.  These ranges
+        # are conservative: the int MUST fall in the range, but the
+        # range may contain ints which may not actually appear in
+        # practice
+        self.var_to_range: Dict["sympy.Symbol", ValueRanges] = {}
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:1096]: It just didn't seem necessary! Also, I view the range analysis on floats as a sort of necessary evil when users do naughty things like sqrt, and for the most part, I only really want to be reasoning about size variables.

File: torch/fx/experimental/symbolic_shapes.py (hunk 1 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:126]: Same deal, not supporting constraints on floats for now

File: torch/utils/_sympy/interp.py [hunk 4]
```diff
@@ -81,8 +81,9 @@ def sympy_interp(
 
     # Recursive case
     args = [sympy_interp(analysis, env, arg) for arg in expr.args]  # type: ignore[arg-type]
-    handler = getattr(analysis, handlers()[expr.func])
-    if handler in ASSOCIATIVE_OPS:
+    handler_name = handlers()[expr.func]
+    handler = getattr(analysis, handler_name)
+    if handler_name in ASSOCIATIVE_OPS:
```
Comment (ezyang) [inline torch/utils/_sympy/interp.py:86]: The old code was just buggy

**Comment by ezyang** [comment]
oops this broke convit_base

**Comment by jeanschmidt** [comment]
@pytorchboot revert -m "Breaking internal builds, more details can be found: https://fburl.com/phabricator/fq5b6k8a" -c nosignal

**Pull request closed without merging**